File: WebScraping/point_alert.py
import os
import sys
import subprocess
from time import sleep
from dotenv import load_dotenv
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.chromium.options import ChromiumOptions

load_dotenv()


class Agent:
    def __init__(self):
        self.link = 'https://web.whatsapp.com/'
        option = ChromiumOptions()
        option.headless = False
        self.driver = webdriver.Chrome(options=option)

    def make_login(self):

        self.driver.get(url=self.link)

        input('Type something to continue: ')

    def enter_conversation(self, contact_name: str):
        span_element = self.driver.find_element(By.XPATH, f"//span[@title='{contact_name}']")
        span_element.click()

    def search_by_specific_message(self, message: str):
        seconds_pass = 0
        while True:
            try:
                self.driver.find_element(By.XPATH, f"//span[.='{message}']")
                print("Element found!")
                self.play_song('acorda_pedrinho.mp3')
                break
            except:
                sleep(1)
                seconds_pass += 1
                print(f'Not Found. Waiting more... Total {seconds_pass} seconds')

        sleep(30)

        self.driver.close()

    @staticmethod
    def play_song(song_path: str):
        if sys.platform.startswith('darwin'):
            # macOS
            subprocess.run(['afplay', song_path])
        elif sys.platform.startswith('win'):
            # Windows
            os.startfile(song_path)
        elif sys.platform.startswith('linux'):
            # Linux
            subprocess.run(['xdg-open', song_path])
        else:
            print('Unsupported platform. Cannot play song.')



contact_name = os.getenv('CONTACT_NAME')

agent = Agent()
agent.make_login()
agent.enter_conversation(contact_name)
agent.search_by_specific_message('.')

File: WebScraping/Node/pointAlert.js
const os = require('os');
const { execSync } = require('child_process');
const { Builder, By } = require('selenium-webdriver');
const chrome = require('selenium-webdriver/chrome');
require('dotenv').config();

const makeLogin = async (link) => {
  const options = new chrome.Options();
  options.headless = false;
  const driver = new Builder().forBrowser('chrome').setChromeOptions(options).build();

  await driver.get(link);
  await new Promise((resolve) => {
    setTimeout(resolve, 20000);
  });

  return driver;
};

const enterConversation = async (driver, contact_name) => {
  const spanElement = await driver.findElement(By.xpath(`//span[@title='${contact_name}']`));
  await spanElement.click();
};

const searchBySpecificMessage = async (driver, message) => {
  let seconds_pass = 0;

  const findElement = async () => {
    try {
      await driver.findElement(By.xpath(`//span[.='${message}']`));
      console.log('Element found!');
      playSong('acorda_pedrinho.mp3');
    } catch {
      seconds_pass += 1;
      console.log(`Not Found. Waiting more... Total ${seconds_pass} seconds`);
      setTimeout(findElement, 1000);
    }
  };

  findElement();

  await new Promise((resolve) => {
    setTimeout(resolve, 30000);
  });

  driver.quit();
};

const playSong = (song_path) => {
  if (os.platform() === 'darwin') {
    // macOS
    execSync(`afplay ${song_path}`);
  } else if (os.platform().startsWith('win')) {
    // Windows
    execSync(`start ${song_path}`);
  } else if (os.platform().startsWith('linux')) {
    // Linux
    execSync(`xdg-open ${song_path}`);
  } else {
    console.log('Unsupported platform. Cannot play song.');
  }
};

(async () => {
  const link = 'https://web.whatsapp.com/';
  const contact_name = process.env.CONTACT_NAME;

  const driver = await makeLogin(link);
  await enterConversation(driver, contact_name);
  await searchBySpecificMessage(driver, '.');
})();
